sklearn/semi_supervised/tests/test_self_training.py

```diff
@@ -0,0 +1,165 @@
+from sklearn.dummy import DummyClassifier
+from sklearn.model_selection import ParameterGrid
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.tree import DecisionTreeClassifier
+from sklearn.svm import SVC
+from sklearn.model_selection import train_test_split
+from sklearn.datasets import load_iris
+from sklearn.utils import check_random_state
+from sklearn.preprocessing import LabelBinarizer
+
+# Author: Oliver Rausch <rauscho@ethz.ch>
+# License: BSD 3 clause
+
+
+rng = check_random_state(0)
+
+# load the iris dataset and randomly permute it
+iris = load_iris()
+perm = rng.permutation(iris.target.size)
+iris.data = iris.data[perm]
+iris.target = iris.target[perm]
+X_train, X_test, y_train, y_test = train_test_split(iris.data,
+                                                    iris.target,
+                                                    random_state=rng)
+
+limit = 50
+
+y_train_missing_labels = y_train.copy()
+y_train_missing_labels[limit:] = -1
+y_train_missing_labels_strings = y_train_missing_labels.copy()
+mapping = {0: 'A', 1: 'B', 2: 'C', -1: -1}
+y_train_missing_strings = np.vectorize(mapping.get)(y_train_missing_labels)
+
+lb = LabelBinarizer()
+y_train_missing_dummies = lb.fit_transform(y_train_missing_labels)
+
+
+def test_classification():
+    # Check classification for various parameter settings.
+    # Also assert that predictions for strings and numerical labels are equal.
+    # Also test for multioutput classification
+    grid = ParameterGrid({"max_iter": [1, 50, 100],
+                          "threshold": [0.0, 0.5, 0.9]})
+
+    for base_classifier in [DummyClassifier(random_state=0),
+                            DecisionTreeClassifier(random_state=0),
+                            KNeighborsClassifier(),
+                            SVC(gamma="scale", probability=True,
+                                random_state=0)]:
+        for params in grid:
+            st = SelfTrainingClassifier(base_classifier, **params)
+            st_string = SelfTrainingClassifier(base_classifier, **params)
+            st.fit(X_train, y_train_missing_labels)
+            pred = st.predict(X_test)
+            proba = st.predict_proba(X_test)
+
+            st_string.fit(X_train, y_train_missing_strings)
+            pred_string = st_string.predict(X_test)
+            proba_string = st_string.predict_proba(X_test)
+
+            assert_array_equal(np.vectorize(mapping.get)(pred), pred_string)
+            assert_array_equal(proba, proba_string)
+
+
+def test_missing_predict_proba():
+    # Check that an error is thrown if predict_proba is not implemented
+    base_classifier = SVC(gamma="scale")
+    self_training = SelfTrainingClassifier(base_classifier)
+    message = "base_classifier (SVC) should implement predict_proba!"
+    assert_raise_message(ValueError, message, self_training.fit, X_train,
+                         y_train)
+
+
+def test_invalid_params():
```

Comment:
I suggested using `@pytest.mark.parametrize`, but if you keep it this way at least avoid using `ParameterGrid` which is overkill here